python/taichi/lang/matrix_ops.py

```diff
@@ -1,12 +1,305 @@
+                if static(inplace):
+                    f(result, m[i, j])
+                else:
+                    result = f(result, m[i, j])
+        return result
+
+    return _reduce()
+
+
+def vector_reduce(v, f, init, inplace=False):
+    shape = v.get_shape()
+
+    @pyfunc
+    def _reduce():
+        result = init
+        for i in static(range(shape[0])):
+            if static(inplace):
+                f(result, v[i])
+            else:
+                result = f(result, v[i])
+        return result
+
+    return _reduce()
+
+
+@preconditions(arg_at(0, assert_tensor))
+def reduce(x, f, init, inplace=False):
+    if len(x.get_shape()) == 1:
+        return vector_reduce(x, f, init, inplace)
+    return matrix_reduce(x, f, init, inplace)
+
+
+@preconditions(
+    arg_at(
+        0,
+        foreach(
+            Or(assert_vector,
+               assert_list,
+               msg="Cols/rows must be a list of lists, or a list of vectors")),
+        same_shapes))
+def rows(rows):  # pylint: disable=W0621
+    if isinstance(rows[0], Matrix):
+        shape = rows[0].get_shape()
+
+        @pyfunc
+        def _rows():
+            return Matrix([[row(i) for i in range(shape[0])] for row in rows])
+
+        return _rows()
+    if isinstance(rows[0], list):
+
+        @pyfunc
+        def _rows():
+            return Matrix([[x for x in row] for row in rows])
+
+        return _rows()
+    # unreachable
+    return None
+
+
+@pyfunc
+def cols(cols):  # pylint: disable=W0621
+    return rows(cols).transpose()
+
+
+def E(m, x, y, n):
+    @func
+    def _E():
+        return m[x % n, y % n]
+
+    return _E()
+
+
+@preconditions(square_matrix,
+               dim_lt(0, 5,
+                      'Determinant of dimension >= 5 is not supported: {}'))
+@func
+def determinant(x):
+    shape = static(x.get_shape())
+    if static(shape[0] == 1 and shape[1] == 1):
+        return x[0, 0]
+    if static(shape[0] == 2 and shape[1] == 2):
+        return x[0, 0] * x[1, 1] - x[0, 1] * x[1, 0]
+    if static(shape[0] == 3 and shape[1] == 3):
+        return x[0, 0] * (x[1, 1] * x[2, 2] - x[2, 1] * x[1, 2]) - x[1, 0] * (
+            x[0, 1] * x[2, 2] - x[2, 1] * x[0, 2]) + x[2, 0] * (
+                x[0, 1] * x[1, 2] - x[1, 1] * x[0, 2])
+    if static(shape[0] == 4 and shape[1] == 4):
+
+        det = 0.0
+        for i in static(range(4)):
+            det += (-1.0)**i * (
+                x[i, 0] *
+                (E(x, i + 1, 1, 4) *
+                 (E(x, i + 2, 2, 4) * E(x, i + 3, 3, 4) -
+                  E(x, i + 3, 2, 4) * E(x, i + 2, 3, 4)) - E(x, i + 2, 1, 4) *
+                 (E(x, i + 1, 2, 4) * E(x, i + 3, 3, 4) -
+                  E(x, i + 3, 2, 4) * E(x, i + 1, 3, 4)) + E(x, i + 3, 1, 4) *
+                 (E(x, i + 1, 2, 4) * E(x, i + 2, 3, 4) -
+                  E(x, i + 2, 2, 4) * E(x, i + 1, 3, 4))))
+        return det
+    # unreachable
+    return None
+
+
+def _vector_transpose(v):
```

Comment:
As we've already handled matrix multiplication with vectors correctly, is it possible to make vector transpose always return a vector?

Reply:
This issue is related to our change here: #6067 
Shall we un-support this operation (we definitely can make it a `v1.outer_product(v2)` though)